libraries/chain/block_handle.cpp

```diff
@@ -13,25 +23,34 @@ void block_handle::write(const std::filesystem::path& state_file) {
    fc::datastream<fc::cfile> f;
    f.set_file_path(state_file);
    f.open("wb");
-
+   fc::raw::pack(f, chain_head_magic);
+   fc::raw::pack(f, chain_head_version);
    fc::raw::pack(f, *this);
 }
 
 bool block_handle::read(const std::filesystem::path& state_file) {
-   if (!std::filesystem::exists(state_file))
-      return false;
+   bool res = false;
+
+   if (!std::filesystem::exists(state_file) || std::filesystem::file_size(state_file) <= 2 * sizeof(chain_head_magic))
+      return res;
 
    fc::datastream<fc::cfile> f;
    f.set_file_path(state_file);
    f.open("rb");
 
-   fc::raw::unpack(f, *this);
+   uint64_t magic, version;
+   fc::raw::unpack(f, magic);
+   fc::raw::unpack(f, version);
 
-   ilog("Loading chain_head block ${bn} ${id}", ("bn", block_num())("id", id()));
+   if (magic == chain_head_magic && version == chain_head_version) {
+      fc::raw::unpack(f, *this);
+      ilog("Loading chain_head block ${bn} ${id}", ("bn", block_num())("id", id()));
+      res = true;
+   }
 
    std::filesystem::remove(state_file);
 
-   return true;
+   return res;
```

Comment:
If the `unpack` fails it throws an exception and the `chain_head.dat` is not removed. If the `chain_head.dat` is less than `2*sizeof(chain_head_magic)` it is not read but the file is not removed. If the magic or version is wrong which it will be for existing 1.0.0 `chain_head.dat`, the file is removed and then the error from `controller::startup` will be `No existing chain_head.dat file`.

Seems we should wrap the whole thing in a try-catch and consistently NOT remove the file. They could downgrade to 1.0.0, and be able to startup. 

I think we should report an error telling user to consider starting from a snapshot if this fails for any reason.

Reply:
done.